source/CCustomOpcodeSystem.cpp

```diff
@@ -1821,19 +1820,108 @@ extern "C"
 		*thread << value;
 	}
 
-	LPSTR WINAPI CLEO_ReadStringOpcodeParam(CLEO::CRunningScript* thread, char *buf, int size)
+	LPCSTR WINAPI CLEO_ReadStringOpcodeParam(CLEO::CRunningScript* thread, char *buf, int bufSize)
 	{
-		return CLEO_ReadStringPointerOpcodeParam(thread, buf, size); // always support all string param types
+		static char internal_buf[MAX_STR_LEN + 1]; // and terminator
+		if (!buf) { buf = internal_buf; bufSize = sizeof(internal_buf); }
+
+		auto result = ReadStringParam(thread, buf, bufSize);
+		if (result == nullptr)
+			LOG_WARNING(thread, "%s in script %s", CCustomOpcodeSystem::lastErrorMsg.c_str(), ((CCustomScript*)thread)->GetInfoStr().c_str());
+
+		return result;
 	}
 
-	LPSTR WINAPI CLEO_ReadStringPointerOpcodeParam(CLEO::CRunningScript* thread, char *buf, int size)
+	LPCSTR WINAPI CLEO_ReadStringPointerOpcodeParam(CLEO::CRunningScript* thread, char *buf, int bufSize)
 	{
-		auto result = ReadStringParam(thread, buf, size);
+		static char internal_buf[MAX_STR_LEN + 1]; // and terminator
+
+		// obtain pointer to the source, null terminated, string data if available
+		char* source = nullptr;
+		auto paramType = thread->PeekDataType();
+		auto arrayType = IsArray(paramType) ? thread->PeekArrayDataType() : eArrayDataType::ADT_NONE;
+		auto isVariableInt = IsVariable(paramType) && (arrayType == eArrayDataType::ADT_NONE || arrayType == eArrayDataType::ADT_INT);
+
+		if (IsImmInteger(paramType) || isVariableInt)
+		{
+			source = (char*)CLEO_PeekIntOpcodeParam(thread); // integer address to text buffer
+
+			if ((size_t)source <= CCustomOpcodeSystem::MinValidAddress)
+			{
+				LOG_WARNING(thread, "Invalid '0x%X' pointer of input string argument #%d in script %s", opcodeParams[0].dwParam, CLEO_GetParamsHandledCount() + 1, ScriptInfoStr(thread).c_str());
+				return nullptr; // error
+			}
+		}

```

Comment:
so previously `CLEO_ReadStringOpcodeParam` supported only string arguments reading them with the native function and returned a pointer to the buf. 

`CLEO_ReadStringPointerOpcodeParam` also supported integer pointers and returned them or string arguments and returned pointer to buffer. in both cases buffer was modified.

Now the native function is re-implemented and supports int pointers too. So there should be no difference between `CLEO_ReadStringOpcodeParam` and `CLEO_ReadStringPointerOpcodeParam`, except different internal buffers.

can we do something like:

```
CLEO_ReadStringOpcodeParam
{
static char internal_buf[MAX_STR_LEN + 1]; // and terminator
if (!buf) { buf = internal_buf; bufSize = sizeof(internal_buf); }
return ReadStringParam(thread, buf, bufSize);
}

CLEO_ReadStringPointerOpcodeParam
{
static char internal_buf[MAX_STR_LEN + 1]; // and terminator
if (!buf) { buf = internal_buf; bufSize = sizeof(internal_buf); }
return ReadStringParam(thread, buf, bufSize);
}

```

?

Reply:
btw, game's native function reading strings is not returning anything useful, the game always reads from the buffer. we can utilize this fact and return original int pointer when integer argument is given, mimicking `CLEO_ReadStringPointerOpcodeParam` behavior

Reply:
Question is can it be changed? Original games function always returned pointer to the provided input buffer. Is it safe to modify it? There is no way to tell who uses/hooks this internal function, better keep it original as caller expect it is his work buffer.

Reply:
> Original games function always returned pointer to the provided input buffer. 

well, not really. here is the code when game reads a fixed 8 byte string:

```
.text:00463D80 014 8B 41 14                                mov     eax, [ecx+CRunningScript.m_pCurrentIP]
.text:00463D83 014 8A 10                                   mov     dl, [eax]
.text:00463D85 014 40                                      inc     eax
.text:00463D86 014 89 41 14                                mov     [ecx+CRunningScript.m_pCurrentIP], eax
.text:00463D89 014 88 16                                   mov     [esi], dl
.text:00463D8B 014 46                                      inc     esi
.text:00463D8C 014 4F                                      dec     edi
.text:00463D8D 014 75 F1                                   jnz     short loc_463D80
.text:00463D8F 014 5F                                      pop     edi
.text:00463D90 010 5E                                      pop     esi
.text:00463D91 00C 5B                                      pop     ebx
.text:00463D92 008 83 C4 08                                add     esp, 8
.text:00463D95 000 C2 08 00                                retn    8
```

you can see that the returned value (eax) will point to SCM location after the end of the string.
I think in original code this function returned `void`, which is why there is no usage of returned value in the code.

Reply:
Also, today the game would simply crash if int is given. So I think it's safe to return `int` value as a result

Reply:
Hm it will make sense then, as all strings processing code will then boil to single function.

Reply:
in the original function, if int argument is given, we still have to copy the content to buffer, so original opcodes will work.
```
03e5: 0xdeadcode // copy 8 bytes from 0xdeadcode to buf
```

in cleo opcodes we can always use result which will be either original string (longer than buf), or buf itself. same logic as CLEO_ReadStringPointerOpcodeParam

Reply:
> Also, today the game would simply crash if int is given. So I think it's safe to return `int` value as a result

Yes, but I implemented support of string pointers into original function on purpose to allow original opcodes use this kind of params. So from game's perspective it should work like with any other input string type.